### ISSUE TITLE
Why does superenv only set $M4 if autoconf needs it to not be?

### ISSUE THREAD

**Issue by gsteemso** [mistydemeo/tigerbrew#1165]
I can't see how anything I did could have affected this, but I have just now started seeing autoconf fail whenever invoked. OK, guy mucks around inside software and it stops working, film at 11, right? Thing is, the cause of the failure is not only something I never went near, but also appears deliberately written to stop it from working in the first place. I know that can't be the case, so what is the logic that I'm not seeing?

Long story short, the environment variable $M4 is always set by superenv if autoconf shows up in the dependencies. When $M4 is set, autoconf uses whatever it points to instead of the one in $PATH – but superenv sets $M4 to the stock system binary, which autoconf can't use! That's why brewed M4 is a dependency of autoconf in the first place. What's the reasoning, there?

**Comment by sevan** [comment]
The requirement for newer m4 than included with darwin is a new change with the release of autoconf at the start of the year.

**Comment by sevan** [comment]
What steps would I need to perform to hit the issue you ran into?

**Comment by gsteemso** [comment]
Honestly, I'm not at all sure.  I was installing packages like normal and all of a sudden, anything that invoked autoconf started failing.

**Comment by gsteemso** [comment]
I've occasionally encountered strange behaviour whenever I had the Leopard “Terminal” app running for a long time (it’s a little more capable but a lot more buggy than the one in Tiger), so I tried rebooting etc.  No change.

**Comment by mistydemeo** [comment]
What does `brew info autoconf` say?

**Comment by gsteemso** [comment]
autoconf: stable 2.72
Automatic configure script builder
https://www.gnu.org/autoconf

This formula is keg-only.
Xcode provides this software prior to version 4.3.

/Users/Shared/Brewery/Cellar/autoconf/2.72 (73 files, 3.7M)
  Built from source
From: https://github.com/mistydemeo/tigerbrew/blob/master/Library/Formula/autoconf.rb
==> Dependencies
Build: xz ✔︎
Required: m4 ✔︎
==> Caveats
This formula is keg-only, which means it was not symlinked into /usr/local.

Xcode provides this software prior to version 4.3.

**Comment by mistydemeo** [comment]
Oh, I see the problem. It's specifically set to call `MacOS.locate`: https://github.com/mistydemeo/tigerbrew/blob/master/Library/Homebrew/extend/ENV/super.rb#L67

But that specifically specifies system-provided copies: https://github.com/mistydemeo/tigerbrew/blob/master/Library/Homebrew/os/mac.rb#L28C5-L33

So it's wrong here. It should be checking if `m4` is in the dep tree, and specifying the Homebrew-installed copy if so.

**Comment by gsteemso** [comment]
That makes more sense, yes. :¬) I guess my biggest question is, "why is $M4 explicitly set in the first place?" It seems a bit random that that program, and no other, gets its own environment variable from superenv. What problem was that introduced to solve?

**Comment by mistydemeo** [comment]
It was a fix for Xcode-only installs on certain OS X versions: https://github.com/Homebrew/legacy-homebrew/issues/24904
